### ISSUE TITLE
Make the mailing list -> PR comment syncing bidirectional

### ISSUE THREAD

**Issue by dscho** [gitgitgadget/gitgitgadget#154]
This will be a very, very tricky thing to implement, with a very real possibility of frequent failures.

We recently saw a couple contributors reply directly in the PR to mails that had been mirrored from the Git mailing list (as PR comments). This does not currently work because the good people on the Git mailing list do not pay attention to GitGitGadget PRs.

To help that, try to figure out which mail to reply to, and send the comment (editing the comment to include a note that it was sent with a link to the mailing list archive).

To figure out what mail to reply to, in the first run, we should look for a quoted link to the mailing list archive (which includes the Message-ID): every PR comment that was mirrored from a corresponding mail on the Git mailing list starts with a link "On the Git mailing list" with a URL to the archive at https://public-inbox.org/git/.

If there is no such quoted link, we could see whether the comment to send contains some quoted text, and if it does not, bail out by adding another comment stating that GitGitGadget could not figure out what mail this reply was for, and please do include the quoted link at the beginning if you want it to be mirrored.

**Comment by dscho** [comment]
See also the thread at https://public-inbox.org/git/20191120142240.pc5kfpj4eflu7dub@yadavpratyush.com/t/#m1bdeef83a936ef244dd1daaef21d34e5bf3b03b4

**Comment by gitster** [comment]
https://github.com/git/git/pull/775 is another example that the contributor found it more convenient to reply in the webform.  The comment by the contributor quotes "On the Git mailing list" (which points to a specific article in the lore archive in the entry added by GGG) but the message-id seems to be redacted---does GGG see the original message-id of the message, or does it only see the redacted form? The solution outlined by the original issue may not work in this case.

**Comment by dscho** [comment]
Indeed. Even if it might be possible *sometimes* to figure out which comment/mail to respond to, in general it won't (e.g. when not quoting anything at all).

In that case, we probably need GitGitGadget to respond with some guidance ("please keep the link to the original mail in the quoted part").

That all sounds quite involved, which is why I did not dare trying to tackle this ticket so far.

**Comment by webstech** [comment]
Mumbling out loud.  There appear to be two different situations.

1. User replying to a GitHub email
2. user adding a PR comment as a reply to a so-and-so@said.this comment by GGG.

In both cases, GGG will now need to process all PR comments.  Not suggesting changes to relieve the poster of reposting/mailing.
1. If it is easily determined this is an email reply (it probably is), a reminder comment can be posted.  May still need to verify a /submit was done.
2. For this case, there is a need to avoid pre-submit comments.  Following the comment chain, if they are not all GGG comments up to the last /submit, the non-GGG comment needs a reminder comment.  This would be limited to comments by the PR owner.

The suggested change is not to 'fix' things but simply remind someone they may not have notified everyone they think they were.

Is this enough?  To put it another way, is this enough for now until we decide the original suggestions about sending emails are needed in the future?

**Comment by dscho** [comment]
Hmm. This would certainly side-step the technical difficulties, but could get pretty annoying pretty quickly, I think: I sometimes post on PRs because it's much more convenient for me. OTOH we could exclude me (and `gitgitgadget`) from that.

**Comment by jonas2515** [comment]
It seems that the main issue here is actually that github's PR ui is not centered around replies to comments, and I guess that can only be worked around by introducing some syntax for that. What sounds doable to me as a compromise would be something like this as comment on the PR:

```
/reply https://github.com/git/git/pull/775#issuecomment-622511614

Thanks I'll look into it!
```

and that would turn into this on the mailing list:
```
From: Jonas Dreßler <verdre@v0yd.nl>

Thanks I'll look into it!
```

for quote replies, there could be
```
/replyquoting https://github.com/git/git/pull/775#issuecomment-622511614

> I do not do P4, so asking those who touched the program in the past
> year and half for opinions.

Thanks I'll look into it!
```

and that would turn into
```
From: Jonas Dreßler <verdre@v0yd.nl>

Junio C Hamano <gitster@pobox.com> writes:

> I do not do P4, so asking those who touched the program in the past
> year and half for opinions.

Thanks I'll look into it!
```

**Comment by jonas2515** [comment]
Tbh I don't really see how specifying which comment you want to reply to is a lot more paintful than having to do `/submit` everytime you push to the PR, a certain level of pain is to be expected if you want to interact with email after all.

The questions of CCing the person sending the reply is tricky, but maybe that's also not necessary if all communication can be done on the github PR? Thinking about it, maybe it's for the best if gitgitgadget wouldn't expose the email of anyone contributing on Github to the mailing list, to not create an incentive for people on the mailing list to reply/write to that persons private email instead of the generated gitgitgadget one.

Fwiw, context is that gitgitgadget looks quite promising for BlueZ and its migration from mailing lists to github, see https://github.com/bluez/bluez/issues/648.

**Comment by dscho** [comment]
> I don't really see how specifying which comment you want to reply to is a lot more paintful than having to do `/submit` everytime you push to the PR

`/replyquoting https://github.com/git/git/pull/775#issuecomment-622511614` is a lot more complex and error-prone than `/submit`.

**Comment by dscho** [comment]
I ran into this again [today](https://github.com/gitgitgadget/git/pull/1757#issuecomment-2211684292). Here are some thoughts how we might move forward with this ticket:

- Often, it is not possible to determine which mail the user is replying to. This is partially a consequence of the relative ease of PR comments vs emailing the reply.
- I _do_ love the ease of PR comments, and I am concerned about the cost of the obstacle that is the Git mailing list (which many claim improves the quality of the contributions, which I consider magical thinking: just because somebody is determined enough to overcome the many requirements to enter the Git project at all does not imply that their contributions will be more in line with what the Git reviewers consider high quality, and vice versa: just because someone is an excellent C programmer does not imply that they are willing to jump the hurdles put in front of them if they want to contribute to Git).
- A quite manual solution to this here ticket would be to introduce the command `/send-reply <URL-to-comment> [<message-ID-or-link-to-mail>]`.
- The implementation of this command would be quite easy: all the building blocks, including the allow list (to determine both who can invoke the `/send-reply` command as well as what comment authors are allowed to be handled in this way), are already in place.
- One caveat is that the mail would need to be in a form that links back to the original comment, and [the automation that handles new mails](https://github.com/gitgitgadget/gitgitgadget/blob/5530f335263bdefccff2f4a6f3fb15459372dc4c/lib/ci-helper.ts#L770-L781) will need to be taught _not_ to mirror them to PR comments.
- It would probably be a good idea for GitGitGadget to edit the original PR comment, adding a link to the mail that was sent (and maybe even a link to the workflow run that sent it).
- From there, we could start figuring out clear indicators in a PR comment that determine conclusively which email to reply to, and make the process less manual by sending the mails automatically whenever that automatic determination successfully found such an `In-Reply-To:` information.
